### PR TITLE
js: add session connection

### PR DESCRIPTION
**Pull request by heljoyLiu**

### CODE REVIEW

**Comment by jorgebay** [comment]
Thanks @heljoyLiu for the patch.

Sorry I didn't have time to review it yet, I'll do it this week.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js [hunk 1]
```diff
@@ -95,11 +95,15 @@ class Connection extends EventEmitter {
     this.isOpen = false;
     this.traversalSource = options.traversalSource || 'g';
     this._authenticator = options.authenticator;
+    this._session = options.session || false;
```
Comment (jorgebay) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js]: We should add `options.session` to the `Connection` and to the `Client` jsdoc.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js [hunk 2]
```diff
@@ -165,6 +169,21 @@ class Connection extends EventEmitter {
...
+        if (bytecode !== null) {
+          reject(new Error('do not support bytecode in session mode'));
+          return;
+        }
+        if (args === null) {
+          reject(new Error('get nil args in request'));
+          return;
+        }
+
+        const message = Buffer.from(this._header + JSON.stringify(this._getSessionRequest(requestId, null, args)))
```
Comment (jorgebay) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js]: Missing `;` at the end of the statement.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js [hunk 3]
```diff
@@ -365,6 +400,11 @@ class Connection extends EventEmitter {
     }
     if (!this._closePromise) {
       this._closePromise = new Promise(resolve => {
+        if (this._session === true) {
+          const message = Buffer.from(this._header + JSON.stringify(this._getSessionRequest(utils.getUuid(), 'close', null)))
```
Comment (jorgebay) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js]: Missing `;`

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js [hunk 4]
```diff
@@ -199,6 +218,22 @@ class Connection extends EventEmitter {
...
+     }
+     args['session'] = this._sessionId;
+ 
+     return ({
+       'requestId': { '@type': 'g:UUID', '@value': id },
+       'op': op || 'eval',
+       'processor': 'session',
+       'args': args
+     });
+   }
```
Comment (jorgebay) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js]: Maybe we could extract common functionality between `_getRequest()` and `_getSessionRequest()`.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js [hunk 5]
```diff
@@ -165,6 +169,21 @@ class Connection extends EventEmitter {
...
 
+      if (this._session === true && op !== 'authentication') {
+        if (bytecode !== null) {
+          reject(new Error('do not support bytecode in session mode'));
+          return;
+        }
+        if (args === null) {
+          reject(new Error('get nil args in request'));
+          return;
+        }
```
Comment (jorgebay) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js]: This should be extracted at the beginning of the method: 

```javascript
submit(bytecode, op, args, requestId, processor) {
  if (bytecode !== null) {
    return Promise.reject(new Error('Bytecode in session mode is not supported'));
  }
  if (args === null) {
    return Promise.reject(new Error('args should not be null'));
  }
}
```

**Comment by spmallette** [comment]
As a quick note @heljoyLiu given that we are trying to quickly turn around a 3.4.6 release to fix a bug in 3.4.5:

https://lists.apache.org/thread.html/r7de7d160173d0774720b347122409bb6e46268613205b08edaa687ee%40%3Cdev.tinkerpop.apache.org%3E

it may take some extra time to see this change merged. 

@jorgebay are you ok with this only going to master branch? should it be aimed at a current release branch instead?

**Comment by jorgebay** [comment]
> are you ok with this only going to master branch? should it be aimed at a current release branch instead?

Thanks for pointing it out.

Would it be possible to base your PR onto `3.3-dev`, @heljoyLiu ?
Otherwise, I can do it myself after review is finished.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js (hunk 1 above)
Comment (heljoyLiu) [reply gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js]: ok, thanks

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js (hunk 2 above)
Comment (heljoyLiu) [reply gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js]: sorry for this obvious mistake

**Comment by heljoyLiu** [comment]
> > are you ok with this only going to master branch? should it be aimed at a current release branch instead?
> 
> Thanks for pointing it out.
> 
> Would it be possible to base your PR onto `3.3-dev`, @heljoyLiu ?
> Otherwise, I can do it myself after review is finished.

ok, Thanks. I will make a new PR with this suggestion on '3.3-dev'

**Comment by heljoyLiu** [comment]
> > are you ok with this only going to master branch? should it be aimed at a current release branch instead?
> 
> Thanks for pointing it out.
> 
> Would it be possible to base your PR onto `3.3-dev`, @heljoyLiu ?
> Otherwise, I can do it myself after review is finished.

I make one new PR on `3.3-dev`,  thanks @jorgebay 

https://github.com/apache/tinkerpop/pull/1251

**Comment by jorgebay** [comment]
Thanks @heljoyLiu , I'll review it shortly.

Feel free to close this one.

**Pull request closed without merging**